Protobuf parser library and first part of working tests.

libretto opened the pull request:
To achieve 1:1 functionality as a SchemaRegistry, we need to use a ProtoBuf schema parser which functionality will be similar to SchemaRegistry used library.  

SchemaRegistry uses a [Square Wire](https://github.com/square/wire) library. It distributed under Apache 2.0 License. But it is written in Kotlin and no analogical library available in python. To use Kotlin code from Karapace we will need JVM as a dependency. To avoid that we decide to port the required ProtoParser code to Python and use it as part of Karapace.  In this pull request, we add ProtoParser code and the first part of ported unittest code for it.

We want to discuss with Aiven representatives which way of using this library is better for the project. May we integrate it now as part of karapace or must we extract from the project and use it as an independent pip library?

hackaugusto commented:
Hey @libretto , thank you for doing this.

>  May we integrate it now as part of karapace or must we extract from the project and use it as an independent pip library?

We had a little chat and starting with the library inside the karapace repo seems like the better option. The idea is that fixing issues is easier when everything is in the same repo.

Again, thanks for doing this!

amrutha-shanbhag commented:
@hackaugusto would be great if you could review and provide feedback if any.

hackaugusto commented:
Hi @amrutha-shanbhag , my colleagues and I will have a look at the PR. It is a big one, so it will take a bit of time before we go over it. Awesome work, thank you :)

juha-aiven commented:
Hey there. A few comments.

I think it'd be worth mentioning the exact source from where the Python code is originating. That way if there are ever fixes / improvements in the original Kotlin code, maybe it'd be possible to port those to Karapace. I think for example `ProtoParserTest` in Wire has been converted to a class with same name in Python. Maybe just add a link back to Wire code in a docstring?

Would it be a biggie to use `pytest` in the tests instead of `unittest`? Karapace uses that elsewhere. I think it'd mostly involve changing `self.assertEqual` to `assert`. For things like CI pipeline it'd be better to have just one test framework.

What comes to the code itself. I think that since the tests have been ported together with the main code, it's very likely things work as expected.

Eventually, we'd write an integration test that runs against Karapace and against Schema Registry and tests protobuf schemas. That'd will reveal if there are any functional differences or problems.

Good work so far! Thanks.

h3nd24 reviewed (commented):
I think this looks really great to me. Just a comment on the unit test. Would it be possible to add more validation errors on each type of element? Mainly to show what happens when certain rules for a particular element is violated.

libretto commented:
@juha-aiven I converted class naming from Kotlin to Python 1:1 (it both are UpperCamelCase) so class ProtoFileElement in Kotlin has the same name in Python. Also, most methods arguments' names and method names themselves were converted from lowerCamelCase in Kotlin to lowercase_separated_by_underscores in Python. Also the same way was converted modules naming. There is the only difference in paths as soon as Kotlin classes are located in different directories. I will add the path to .kt file in the .py file for reference. 

I selected unittest because is similar to junit by asserts and it adds some comfort and performance to porting process. But now when the test porting process is fast enough and fewer errors appear I can convert it to pytest at once...

mrlika commented:
> OK, sounds good. So, what's remaining to finish the feature is:
> 
>     1. Port the remaining tests from Kotlin to Python
> 
>     2. Change `unittest` to `pytest`
> 
>     3. Implement what's needed on the Karapace.
> 
>     4. Add integration tests
> 
>     5. Verify compatibility with Schema Registry
> 
> 
> The steps 3-5 are kind of simultaneous.
> 
> Anything else?

Everything correct. Change `unittest` to `pytest` and mentioning the exact source will be done in this PR.

We are going to focus on `Implement what's needed on the Karapace.` after that. It will be implemented in the next PRs.

Then Instaclustr team will work on 1, 4 while Karapace team can start verifying compatibility (5).

juha-aiven commented:
> > Anything else?
> 
> Everything correct. Change `unittest` to `pytest` and mentioning the exact source will be done in this PR.
> 
> We are going to focus on `Implement what's needed on the Karapace.` after that. It will be implemented in the next PRs.
> 
> Then Instaclustr team will work on 1, 4 while Karapace team can start verifying compatibility (5).

Sounds like a plan. You can change this PR to a draft and mark it for review when it's time to merge it in.

libretto commented:
@juha-aiven I changed test framework to pytest and added references to the square/wire. I think this PR coding is finished. And next part of tests, library utilization e.t.c will be in other PRs.

juha-aiven commented:
Good. There's now `pytest`.

Could you open a PR in https://github.com/aiven/karapace? Now the #6 is already merged and this PR builds on top of that. Seeing the overall diff with https://github.com/aiven/karapace master is a little hard.

I think we should in https://github.com/aiven/karapace have first a PR that pulls in the stuff in this PR, essentially the protobuf parsing and the associated tests. At this point we wouldn't expose Protobu schema support in the API.

After that PR is merged, the next PR would add the Protobuf support to Karapace. That PR should also contain integration tests. We don't want to go live with Protobuf support without good integration test coverage.

How does this sound?

juha-aiven commented:
The other approach is that the feature is implemented fully in this fork (with integration tests and all) and merged then to https://github.com/aiven/karapace? So far it looks like this is the approach taken. This works for us too.

mrlika commented:
> The other approach is that the feature is implemented fully in this fork (with integration tests and all) and merged then to https://github.com/aiven/karapace? So far it looks like this is the approach taken. This works for us too.

We are going to implement the full feature in this fork. We keep the fork in sync with `aiven/karapace`.

libretto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,380 @@
+    return -1
+
+
+def min_of(a: int, b: int) -> int:
+    return a if a < b else b
```
ok

libretto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,380 @@
+from karapace.protobuf.location import Location
+
+
+def hex_digit(c: str) -> int:
+    if ord(c) in range(ord('0'), ord('9') + 1):
+        return ord(c) - ord('0')
+    if ord(c) in range(ord('a'), ord('f') + 1):
+        return ord(c) - ord('a') + 10
+    if ord(c) in range(ord('A'), ord('F') + 1):
+        return ord(c) - ord('A') + 10
+    return -1
```
ok

libretto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,380 @@
+
+        self.unexpected("unterminated string")
+
+    def read_numeric_escape(self, radix: int, length: int) -> str:
+        value = -1
+        end_pos = min_of(self.pos + length, len(self.data))
+        while self.pos < end_pos:
+            digit = hex_digit(self.data[self.pos])
+            if digit == -1 or digit >= radix:
+                break
+
+            if value < 0:
+                value = digit
+            else:
+                value = value * radix + digit
+            self.pos += 1
+
+        self.expect(value >= 0, "expected a digit after \\x or \\X")
+        return chr(value)
```
This code is part of parsing library. Number of digits is unknown so we cannot use int() call there as soon as we will receive exception.

libretto replied:
About encoding - I cannot find info about affect of encoding on python call chr( digits ).

hackaugusto replied:
> Number of digits is unknown so we cannot use int() call there as soon as we will receive exception.

Isn't that what the arg `length' does?

> In both cases it is unicode representation of least 16 bit of int value.

cool, thanks for checking.

libretto replied:
Length limits number of chars which must be parsed But in case if non digit met then routine return number which found or exception if no number found. So 0x1\t must return 0x1 but int("0x1\t") return exception. 

syntax_reader.py contain code fully ported from square/wire using their parsing method. It is reason why there no regexps. Because they absent in Kotlin code too. I think optimizing and improving of this part  must be done only in case of performance issues and benchmarks.

libretto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,380 @@
+                tag = tag[len("0x"):]
+                radix = 16
+            return int(tag, radix)
+        except OSError as err:
+            print("OS error: {0}".format(err))
```
I removed it...

libretto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,380 @@
+        try:
+            radix = 10
+            if tag.startswith("0x") or tag.startswith("0X"):
+                tag = tag[len("0x"):]
```
yes it maybe removed.
  tag = tag[len("0x"):]

libretto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,380 @@
+                self.expect(self.pos < len(self.data), "unexpected end of file")
+                c = self.data[self.pos]
+                self.pos += 1
+                d: str = {
+                    'a': "\u0007",  # Alert.
+                    'b': "\b",  # Backspace.
+                    'f': "\u000c",  # Form feed.
+                    'n': "\n",  # Newline.
+                    'r': "\r",  # Carriage return.
+                    't': "\t",  # Horizontal tab.
+                    'v': "\u000b",  # Vertical tab.
+                }.get(c)
```
This code is part of ported parser from square/wire library. They do not use regexp for parsing...  Code is consistent with other logic. So question is: will it be faster use regexps to only find string inside quotes?

hackaugusto replied:
> So question is: will it be faster use regexps to only find string inside quotes?

I'm not sure, would need to do [some benchmarks](docs.python.org/3/library/timeit). I was more worried about missing valid escape sequences.

libretto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,380 @@
+            c = self.data[self.pos]
+            self.pos += 1
+            if c == start_quote:
+                if self.peek_char() == '"' or self.peek_char() == "'":
```
this logic ported from square/wire. and there no skip_whitespace() call

libretto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,380 @@
+        if c == '(':
+            self.pos += 1
+            result = self.read_word()
+            self.expect(self.read_char() == ')', "expected ')'")
```
Yes we can replace it  but we will not do that:
1) I think better idea is  - get rid of self.require() method at all as soon ass it is  wrapper without additional functionality. 
2) This code is port of square/wire and it is close to original code in Kotlin. So I prefer keep it as is.

libretto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,380 @@
+        if c == '[':
+            self.pos += 1
+            result = self.read_word()
+            self.expect(self.read_char() == ']', "expected ']'")
```
up

libretto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,380 @@
+    def read_data_type_by_name(self, name: str) -> str:
+        """ Reads a scalar, map, or type name with `name` as a prefix word. """
+        if name == "map":
+            self.expect(self.read_char() == '<', "expected '<'")
```
answered above

libretto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,380 @@
+            self.expect(self.read_char() == '<', "expected '<'")
+            key_type = self.read_data_type()
+
+            self.expect(self.read_char() == ',', "expected ','")
```
answered above

libretto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,380 @@
+            self.expect(self.read_char() == ',', "expected ','")
+            value_type = self.read_data_type()
+
+            self.expect(self.read_char() == '>', "expected '>'")
```
answered above

libretto commented on karapace/protobuf/utils.py:
```diff
@@ -1,3 +1,66 @@
+#        return (MIN_TAG_VALUE <= self <= RESERVED_TAG_VALUE_START) or\
+#               (RESERVED_TAG_VALUE_END + 1 <= self <= MAX_TAG_VALUE + 1)
+
+# builtins.int = MyInt
```
It was a workaround when i worked on porting and I forget to remove it...

libretto commented on karapace/protobuf/syntax_reader.py:
```diff
@@ -0,0 +1,380 @@
+        """ Reads a non-empty word and returns it. """
+        self.skip_whitespace(True)
+        start = self.pos
+        while self.pos < len(self.data):
+            c = self.data[self.pos]
+            if ord(c) in range(ord('a'), ord('z') + 1) \
+                    or ord(c) in range(ord('A'), ord('Z') + 1) \
+                    or ord(c) in range(ord('0'), ord('9') + 1) or c in ['_', '-', '.']:
+                self.pos += 1
+            else:
+                break
```
It is direct port of square wire kotlin code. I prefer to keep without regexp.

libretto commented on karapace/protobuf/utils.py:
```diff
@@ -1,3 +1,66 @@
+RESERVED_TAG_VALUE_END = 19999
+""" True if the supplied value is in the valid tag range and not reserved.  """
+
+# class MyInt(int):
```
removed already

libretto commented on karapace/protobuf/type_element.py:
```diff
@@ -0,0 +1,23 @@
+from karapace.protobuf.location import Location
+
+
+class TypeElement:
```
ok
